### PR TITLE
[ExportSystemC] Support operations from func dialect

### PR DESCRIPTION
**Pull request by maerhart**
Emit functions from the func dialect as regular C functions. This is useful for interop when we, e.g., want to wrap some C++ code in a C function such that a caller which does not support C++ can still call that code.

`call_indirect` is also especially useful for calling member functions accessed by `systemc.cpp.member_access`.

### CODE REVIEW

File: lib/Target/ExportSystemC/Patterns/BuiltinEmissionPatterns.cpp [hunk 1]
```diff
@@ -76,6 +76,36 @@ struct IndexTypeEmitter : TypeEmissionPattern<IndexType> {
...
+    p << "(";
+
+    bool first = true;
+    for (Type ty : type.getInputs()) {
+      if (!first)
+        p << ", ";
+
+      p.emitType(ty);
+      first = false;
+    }
```
Comment (uenoku) [inline lib/Target/ExportSystemC/Patterns/BuiltinEmissionPatterns.cpp:103]: Probably you can simplify this with `llvm::interleaveComma` 
```suggestion
   llvm::interleaveComma(type.getInputs(), p,
                          [&](auto ty) { p.emitType(ty); });
```

File: lib/Target/ExportSystemC/EmissionPrinter.cpp [hunk 2]
```diff
@@ -89,7 +89,7 @@ InlineEmitter EmissionPrinter::getInlinable(Value value) {
       << value << "'\n";
   err.attachNote(requestLoc) << "requested to be inlined here";
   return InlineEmitter(
-      [&]() { os << "<<INVALID VALUE TO INLINE (" << value << ")>>"; },
+      [=]() { os << "<<INVALID VALUE TO INLINE (" << value << ")>>"; },
```
Comment (seldridge) [inline lib/Target/ExportSystemC/EmissionPrinter.cpp:92]: Why is this changed to capture by copy?

File: lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp [hunk 3]
```diff
@@ -0,0 +1,252 @@
...
+namespace {
+/// Emit a func.func function. Users of the function arguments request an
+/// expression to be inlined and we simply return the name of the argument.This
+/// name has to be passed to this emission pattern via an array of strings
+/// attribute called 'argNames' because the emitter cannot do any name uniquing
+/// as it just emits the IR statement by statement. However, relying on an
+/// attribute for the argument names also has the advantage that the names
+/// can be preserved during a lowering pipeline and upstream passes have more
+/// control on how the arguments should be named (e.g. when they create a
+/// function and have some context to assign better names).
```
Comment (seldridge) [inline lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp:34]: The `argNames` seems like a necessary evil. I don't see a good way around this without requiring changes to upstream `func` dialect. The intent here is that names will need to be legalized via a pass before the conversion is run, yes?

File: lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+          return Precedence::VAR;
+      }
+    }
+
+    return {};
+  }
+
+  void emitInlined(Value value, EmissionPrinter &p) override {
+    auto func = value.getParentRegion()->getParentOfType<FuncOp>();
+    for (size_t i = 0, e = func.getNumArguments(); i < e; ++i) {
```
Comment (seldridge) [inline lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp:55]: Consider a zip of arguments and `argNames`.

File: lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+
+      if (funcOp.getNumArguments() == 0)
+        return true;
+
+      if (auto argNames = funcOp->getAttrOfType<ArrayAttr>("argNames"))
+        return argNames.size() == funcOp.getNumArguments() &&
+               llvm::all_of(argNames, [](Attribute arg) {
+                 return arg.isa<StringAttr>();
+               });
+    }
```
Comment (seldridge) [inline lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp:79]: This is kind of spooky. This is only able to emit func ops which meet a certain, higher standard. This seems slightly concerning as its overlaying a new level of verification on top of func ops which has a strange failure mode for users: if they don't properly populate a valid `argNames` array, then the emission just won't work.

I don't know how to handle this. 😕 This may hint that emitting `funcOp` is not the right way to go and we want something else that `funcOp` is converted to. That can be safely verified with things like `argNames` existing, being the right type, and having the right size.

File: test/Target/ExportSystemC/errors.mlir [hunk 6]
```diff
@@ -9,3 +9,48 @@ hw.module @notSupported () -> () { }
 // CHECK: <<UNSUPPORTED TYPE (!hw.inout<i2>)>>
 // expected-error @+1 {{no emission pattern found for type '!hw.inout<i2>'}}
 systemc.module @invalidType (%port0: !systemc.in<!hw.inout<i2>>) {}
+
+// -----
+
+// expected-error @+1 {{no emission pattern found for type '() -> (i32, i32)'}}
+func.func private @functionTypeMultipleResults (%a: () -> (i32, i32)) attributes {argNames=["a"]}
```
Comment (seldridge) [inline test/Target/ExportSystemC/errors.mlir:16]: This type of error isn't awesome. The pass writer is going to wonder why this failed. Ditto for other things in this test. (Noting, that I love the exhaustiveness of this test! I'm just wondering if there is a better way of checking this and giving feedback to the pass writer.)

**Review by seldridge — COMMENTED** [review]
Overlaying the `argNames` requirement on top of `func.func` is kind of spooky. Are there other approaches that we could take to accomplishing the same thing? E.g., maybe we want a separate op that can be verified based on the requirements of emission.

File: lib/Target/ExportSystemC/EmissionPrinter.cpp (hunk 2 above)
Comment (maerhart) [reply lib/Target/ExportSystemC/EmissionPrinter.cpp:92]: See the GCC build failure above. There was an issue with variadic result values such as `%0:3 = call ...`. I don't know exactly how this is handled in MLIR, but basically the reference to `value` was invalid at the point where this lambda is called.

File: lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp (hunk 3 above)
Comment (maerhart) [reply lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp:34]: Correct (assuming conversion refers to emission).

File: test/Target/ExportSystemC/errors.mlir (hunk 6 above)
Comment (maerhart) [reply test/Target/ExportSystemC/errors.mlir:16]: I agree that it's hard to find out what the cause of this error is. It's the same issue that the dialect conversion framework has. When you implement a set of patterns there and you made some mistake or your pattern does not match on a case you thought could not exist, you get an error stating that the legalization of a certain operation failed without stating what patterns it matched against and why each of them failed. Compared to the dialect conversion framework we don't  have a `--debug-only=dialect-conversion` kind of debug log which is very helpful in such cases. Of course, it would be possible to also implement something like this for this printer. It would also be possible to collect messages on the way on why each pattern did not match and then listing them in the final error message, but we would need to make sure that only useful messages are attached to it (not something like "did not match because the pattern only accepts `func.func` ops and it got a `func.call` op"). So the reason why this error message it not very nice is not that it is impossible, but that a proper implementation requires more than a few lines of code (requires a separate PR) and that I didn't prioritize it yet.

File: lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp (hunk 5 above)
Comment (maerhart) [reply lib/Target/ExportSystemC/Patterns/FuncEmissionPatterns.cpp:79]: I agree that it is not very elegant (quite hacky actually) to just add attributes that are important for correctness of emission to an op that has no native support for it. Trying to add the features we need to the upstream `func.func` is no option since some things we need don't fit the upstream dialect. I'd suggest to create our own custom function operation that can encode all the features we need natively (as ODS attributes, etc.), namely
* argNames
* extern "C" support
* verifying that there is only one return value as in C/C++ functions
* etc.

I wouldn't call additional matching conditions "concerning as its overlaying a new level of verification" though. It's weird here because they check for some seemingly random attributes, but having several separate emission patterns for the same op is totally fine (e.g., one for functions with 0 return values, one for functions with 1 return value, one for functions with more than 1 return values) Although, I would recommend handling the latter case in a pre-pass and reduce it to one return value rather than dealing with it in the emitter. In that case it would just check that the pre-pass was executed and did all of its work and it's fine to fail if not. Aiming to cover all cases in the printer directly will just lead to one big monolithic blob of pass that's hard to understand, maintain, and reuse. But converting to a new operation just dedicated to represent (and verify) the subset of another operation supported by the printer (sitting at the end of the emission pipeline) may not be the best approach in all cases. In that sense, yes, the matching function of the pattern does some form of verification to check that the rest of the emission pipeline did it's work and nothing was forgotten there, but it's likely better than just introducing a new op to move that kind of verification to or to just crash. Sorry for the rambling 😅

**Review by seldridge — APPROVED** [review]
Looks reasonable.

I'm nervous about the duplication from `func` and long term maintainability. However, as we've discussed, this appears to be the strategy in upstream dialects with similar problems.

File: include/circt/Dialect/SystemC/SystemCStructure.td [hunk 7]
```diff
@@ -207,3 +207,167 @@ def DestructorOp : SystemCOp<"cpp.destructor", [
   let hasVerifier = true;
   let skipDefaultBuilders = 1;
 }
+
+def FuncOp : SystemCOp<"cpp.func", [
```
[the remark this answers is not in the document]
Comment (maerhart) [reply include/circt/Dialect/SystemC/SystemCStructure.td:211]: I completely agree with you. It would be great to have a way in TableGen to inherit from an operation (creating no sub-type relation or such, but just to reuse the code) to add a few attributes and only have to deal with verification, etc. of those. Or alternatively have utilities in TableGen to easily create specialized FuncOps with their return and call operations (i.e. tablegen automatically generates the verification code to check the operands of those three ops match in terms of size and types, etc.).
